File: pkmnapi-api/src/requests/pokemon_icons.rs
use rocket_okapi::JsonSchema;
use serde::Deserialize;

use crate::requests::base::BaseRequest;

pub type PokemonIconRequest = BaseRequest<PokemonIconRequestType, PokemonIconRequestAttributes>;

impl PokemonIconRequest {
    pub fn get_icon_id(&self) -> u8 {
        self.data.attributes.icon.id
    }
}

#[derive(Debug, Deserialize, JsonSchema)]
#[allow(non_camel_case_types)]
pub enum PokemonIconRequestType {
    pokemon_icons,
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct PokemonIconRequestAttributes {
    pub icon: PokemonIconRequestAttributesIcon,
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct PokemonIconRequestAttributesIcon {
    #[serde(deserialize_with = "crate::utils::from_numeric_str")]
    pub id: u8,
}

File: clients/js/src/requests/pokemon_icons.ts
import { BaseRequest } from './base';

export type PokemonIconRequest = BaseRequest<PokemonIconRequestType, PokemonIconRequestAttributes>;

enum PokemonIconRequestType {
    pokemon_icons = 'pokemon_icons',
}

export interface PokemonIconRequestAttributes {
    icon: PokemonIconRequestAttributesIcon;
}

export interface PokemonIconRequestAttributesIcon {
    id: string;
}

export const PokemonIconRequestDefault = (attributes: PokemonIconRequestAttributes): PokemonIconRequest => {
    return {
        data: {
            type: PokemonIconRequestType.pokemon_icons,
            attributes,
        },
    };
};
